File: agent-cpp/src/logic.cc
#include <fmt/ranges.h>
#include <spdlog/spdlog.h>

#include <vector>

#include "agent/agent.h"
#include "agent/position.h"
#include "path_finding.h"
#include "shot_check.h"

constexpr auto kFloatPositionShift = 0.5;

// NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables)
static struct {
  std::vector<thuai7_agent::Position<int>> path;
} state;
// NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables)

void Setup(thuai7_agent::Agent& agent) {
  // Your code here.
}

void Loop(thuai7_agent::Agent& agent) {
  // Your code here.
  // Here is an example of how to use the agent.
  // Always move to the opponent's position, keep one cell away from the
  // opponent, and attack the opponent.

  auto const& player_info_list = agent.all_player_info().value().get();
  auto const self_id = agent.self_id().value();
  auto const& self_info = player_info_list.at(self_id);
  auto const& opponent_info =
      player_info_list.at((self_info.id + 1) % player_info_list.size());

  auto const& map = agent.map().value().get();

  auto self_position_int =
      thuai7_agent::Position<int>{static_cast<int>(self_info.position.x),
                                  static_cast<int>(self_info.position.y)};
  auto opponent_position_int =
      thuai7_agent::Position<int>{static_cast<int>(opponent_info.position.x),
                                  static_cast<int>(opponent_info.position.y)};

  if (std::find(state.path.begin(), state.path.end(), self_position_int) ==
          state.path.end() ||
      std::find(state.path.begin(), state.path.end(), opponent_position_int) ==
          state.path.end()) {
    state.path = FindPathBeFS(map, self_position_int, opponent_position_int);

    if (state.path.empty()) {
      spdlog::info("no path from {} to {}", self_position_int,
                   opponent_position_int);
      return;
    }

    spdlog::info("found path from {} to {}", self_position_int,
                 opponent_position_int);
  }

  while (state.path.front() != self_position_int) {
    state.path.erase(state.path.begin());
  }

  if (state.path.size() > 1) {
    auto next_position_int = state.path.at(1);
    auto next_position = thuai7_agent::Position<float>{
        static_cast<float>(next_position_int.x + kFloatPositionShift),
        static_cast<float>(next_position_int.y + kFloatPositionShift)};

    agent.Move(next_position);
    return;
  }

  if (CheckShotFeasible(map, self_info.position, opponent_info.position,
                        self_info.range)) {
    agent.Attack(opponent_info.position);
  }
}

File: agent-python/logic.py
import logging
from typing import List, Optional

from pathfinding.core.grid import Grid
from pathfinding.finder.best_first import BestFirst

from agent.agent import Agent
from agent.position import Position

game_map_matrix: Optional[List[List[int]]] = None
path: List[Position[int]] = []


async def setup(agent: Agent):
    # Your code here.
    pass


async def loop(agent: Agent):
    # Your code here.
    # Here is an example of how to use the agent.
    # Always move to the opponent's position, keep one cell away from the
    # opponent, and attack the opponent.

    player_info_list = agent.all_player_info
    assert player_info_list is not None

    self_id = agent.self_id
    assert self_id is not None

    self_info = player_info_list[self_id]
    opponent_info = player_info_list[(self_id + 1) % len(player_info_list)]

    game_map = agent.map
    assert game_map is not None

    global game_map_matrix

    if game_map_matrix is None:
        game_map_matrix = [
            [1 for _ in range(game_map.length)] for _ in range(game_map.length)
        ]
        for obstacle in game_map.obstacles:
            game_map_matrix[obstacle.y][obstacle.x] = 0

    self_position_int = Position[int](
        int(self_info.position.x), int(self_info.position.y)
    )
    opponent_position_int = Position[int](
        int(opponent_info.position.x), int(opponent_info.position.y)
    )

    global path

    if self_position_int not in path or opponent_position_int not in path:
        path = find_path_befs(game_map_matrix, self_position_int, opponent_position_int)

        if len(path) == 0:
            logging.info(
                "no path from %s to %s", self_position_int, opponent_position_int
            )
            return

        logging.info(f"found path from {self_position_int} to {opponent_position_int}")

    while path[0] != self_position_int:
        path.pop(0)

    if len(path) > 1:
        next_position_int = path[1]
        next_position = Position[float](
            float(next_position_int.x) + 0.5, float(next_position_int.y) + 0.5
        )

        await agent.move(next_position)
        return

    await agent.attack(opponent_info.position)


def find_path_befs(
    game_map_matrix: List[List[int]], start: Position[int], end: Position[int]
) -> List[Position[int]]:
    game_map_grid = Grid(matrix=game_map_matrix)
    start_node = game_map_grid.node(start.x, start.y)
    end_node = game_map_grid.node(end.x, end.y)
    finder = BestFirst()
    path, _ = finder.find_path(start_node, end_node, game_map_grid)
    assert isinstance(path, list)
    return [Position[int](x, y) for x, y in path]
